src/category_theory/sites/sheaf.lean

```diff
@@ -0,0 +1,830 @@
+  begin
+    rintro ⟨x, hx⟩,
+    refl,
+  end }
+
+/-- (Implementation). A lemma useful to prove `yoneda_condition_iff_sheaf_condition`. -/
+lemma extension_iff_amalgamation (x : S.functor ⟶ P) (g : yoneda.obj X ⟶ P) :
+  S.functor_inclusion ≫ g = x ↔
+  (nat_trans_equiv_compatible_family x).1.is_amalgamation (yoneda_equiv g) :=
+begin
+  change _ ↔ ∀ ⦃Y : C⦄ (f : Y ⟶ X) (h : S f), P.map f.op (yoneda_equiv g) = x.app (op Y) ⟨f, h⟩,
+  split,
+  { rintro rfl Y f hf,
+    rw yoneda_equiv_naturality,
+    dsimp,
+    simp },
+  { intro h,
+    ext Y ⟨f, hf⟩,
+    have : _ = x.app Y _ := h f hf,
+    rw yoneda_equiv_naturality at this,
+    rw ← this,
+    dsimp,
+    simp },
+end
+
+/--
+The yoneda version of the sheaf condition is equivalent to the sheaf condition.
+
+C2.1.4 of [Elephant].
+-/
+lemma yoneda_condition_iff_sheaf_condition :
+  is_sheaf_for P S ↔ yoneda_sheaf_condition P S :=
+begin
+  rw [is_sheaf_for, yoneda_sheaf_condition],
+  simp_rw [extension_iff_amalgamation],
+  rw equiv.forall_congr_left' nat_trans_equiv_compatible_family,
+  rw subtype.forall,
+  apply ball_congr,
+  intros x hx,
+  rw equiv.exists_unique_congr_left _,
+  simp,
+end
+
+/-- `P` is a sheaf for `R` iff it is separated for `R` and there exists an amalgamation. -/
+lemma is_separated_for_and_exists_is_amalgamation_iff_sheaf_for :
+  is_separated_for P R ∧ (∀ (x : family_of_elements P R), x.compatible → ∃ t, x.is_amalgamation t) ↔
+  is_sheaf_for P R :=
+begin
+  rw [is_separated_for, ←forall_and_distrib],
+  apply forall_congr,
+  intro x,
+  split,
+  { intros z hx, exact exists_unique_of_exists_of_unique (z.2 hx) z.1 },
+  { intros h,
+    refine ⟨_, (exists_of_exists_unique ∘ h)⟩,
+    intros t₁ t₂ ht₁ ht₂,
+    apply (h _).unique ht₁ ht₂,
+    exact is_compatible_of_exists_amalgamation x ⟨_, ht₂⟩ }
+end
+
+/-- If `P` is separated for `R` and every family has an amalgamation is a sheaf. -/
+lemma is_separated_for.is_sheaf_for (t : is_separated_for P R) :
+  (∀ (x : family_of_elements P R), x.compatible → ∃ t, x.is_amalgamation t) →
+  is_sheaf_for P R :=
+begin
+  rw ← is_separated_for_and_exists_is_amalgamation_iff_sheaf_for,
+  apply and.intro t,
+end
+
+/-- Get the amalgamation of the given compatible family, provided we have a sheaf. -/
+noncomputable def is_sheaf_for.amalgamate
+  (t : is_sheaf_for P R) (x : family_of_elements P R) (hx : x.compatible) :
+  P.obj (op X) :=
+classical.some (t x hx).exists
+
+lemma is_sheaf_for.is_amalgamation
+  (t : is_sheaf_for P R) {x : family_of_elements P R} (hx : x.compatible) :
+  x.is_amalgamation (t.amalgamate x hx) :=
+classical.some_spec (t x hx).exists
+
+@[simp]
+lemma is_sheaf_for.valid_glue
+  (t : is_sheaf_for P R) {x : family_of_elements P R} (hx : x.compatible) (f : Y ⟶ X) (Hf : R f) :
+  P.map f.op (t.amalgamate x hx) = x f Hf :=
+t.is_amalgamation hx f Hf
+
+/-- If `P` is a sheaf for `R`, it is separated for `R`. -/
+lemma is_sheaf_for.is_separated_for : is_sheaf_for P R → is_separated_for P R :=
```

Comment:
Why not put this directly after `is_separated_for.is_sheaf_for`? Feels more like where it belongs.

Reply:
Good point, thanks!